### ISSUE TITLE
Registration of Twitter user accounts

### ISSUE THREAD

**Issue by Orbiter** [loklak/loklak_webclient#38]
loklak shall be authorized to act as twitter application. This shall enable loklak to tweet in the users name. Possible applications are
- create statistics and tweet these (automatically or manually)
- cross-tweeting (i.e. writing messages within the loklak environment and loklak pushes them to twitter)
- producing 'enriched' tweets (i.e. with attached better location coordinates, more text, more embedded media - we will find a way to 'compress' such things into cross-tweets)

Thats just the current vision, not the requirement for this feature!

For this step, we just need
- a "sign-in" page
- authentify the entered account credentials against twitter
- store the credentials 'somewhere' (where? if necessary, hand the credentials over to the loklak back-end server, I will provide a storage and retrieval api, if necessary)

**Comment by Orbiter** [comment]
Assigned this to Julian because Mario told me that Julian said that this is easy to do :blush: 

@julianfox
- currently the loklak_webclient can be hosted statically, does an account verification need a running node.js environment? Or can this also be done in the loklak_server?
- where are the account credentials stored? If a node.js environment stores that, what kind of technology is doing that?
- Can we enable the back-end to use the account credentials as well? I believe this is necessary because some content for cross-posting is generated there.

**Comment by julianfox** [comment]
Hey,
The best way to do it for me is to use a token based auth with a lib like https://github.com/sahat/satellizer
You do not need to use node.js, we can manage it with the loklak_server.

Satellizer provide a good java exemple for the server side: https://github.com/sahat/satellizer/tree/master/examples/server/java

In this way, the account credentials is stored in db managed by the server side so the back-end can use the account credentials.

For be honest, i never use it but i already did a benchmark for the next version of HV.

**Comment by Orbiter** [comment]
I had a look at satellizer; looks like a small library. Maybe I can replace the database they use elasticsearch and a text dump (for a backup).

**Comment by julianfox** [comment]
Yes of course, i think you can use what you want, but i never used Satellizer yet,

**Comment by mariobehling** [comment]
@julianfox So, how can we get started here? Could you help to implement a first version?

**Comment by julianfox** [comment]
Yes of course, i am working on the wall right now, may be this feature can be the next step ? This have to be implemented before the openTech ?

**Comment by mariobehling** [comment]
It would be nice to implement it before the OpenTechSummit, but the twitter wall is the most important.

**Comment by Orbiter** [comment]
lets do that now: Julian, whats necessary for the server back-end? What shall it store. How do we organize sessions once a user has logged in?

**Comment by julianfox** [comment]
I thinks that the best solution is to use Satellizer as i said, so for the server back-end, we can use the actual back-end and reproduce this: https://github.com/sahat/satellizer/tree/master/examples/server/java

What do you think about this solution ?
If it's ok on your side, i can implement the front-end side

**Comment by julianfox** [comment]
Here a complete description of the workflow: https://github.com/sahat/satellizer/wiki/Login-with-OAuth-2.0

**Comment by julianfox** [comment]
@sudheesh001 you wrote about passportjs in your last daily scrum, what the pros and cons for you about Satellizer vs Passportjs ?

Can we share data between the front-end and the back-end with Passportjs ?

**Comment by sudheesh001** [comment]
@julianfox I've never used either, I'll experiment with them and report back.

**Comment by sudheesh001** [comment]
@julianfox I've started a separate fork of this web_client in Python using flask, it'll be functioning exactly the same way as our current repo is with the same functionality within the next few commits, it should be easier to do the DB and the twitter integrations from there, how about we try that and pitch it to rebase this project and continue, we're still starting off so changing the technology shouldn't be a problem.

**Comment by Orbiter** [comment]
sounds good

**Comment by aneeshd16** [comment]
Twitter auth will also be used by the wall, so that users can login, create and store different twitter walls. Please create the db and auth to accommodate the wall as well.

**Comment by julianfox** [comment]
ok perfect, go for it

**Comment by Orbiter** [comment]
The loklak_server now provides storage of authentication. If you run your own loklak_server, you can call i.e.
```
http://localhost:9100/api/account.json?action=update&data={%22screen_name%22:%22myaccount%22,%22consumerKey%22:%221x3%22,%22consumerSecret%22:456,%22source_type%22:%22TWITTER%22}
```
...to push a consumer key and secret.
To retrieve that information, call
```
http://localhost:9100/api/account.json?screen_name=myaccount
```

**Comment by Orbiter** [comment]
according to current discussion, the storage of consumer keys is done statically only once and the backend stores the oauth tokens. This was now changed with https://github.com/loklak/loklak_server/commit/4f2b31ea5f8d524332edfb8c9b267cb1f670090e
New calls are:

write:
```
http://localhost:9100/api/account.json?action=update&data={%22screen_name%22:%22myaccount%22,%22oauth_token%22:%221x3%22,%22oauth_token_secret%22:456,%22source_type%22:%22TWITTER%22}
```
read:
```
http://localhost:9100/api/account.json?screen_name=myaccount
```

**Comment by sudheesh001** [comment]
Can be closed after trials patching https://github.com/loklak/loklak_server/issues/43 
Currently the auth maintains sessions and throws error
`The user was not saved in loklak_server. Handle this error` which is present at https://github.com/loklak/loklak_webclient/blob/76eccf6a7c6a17cadc67df98ad89e4cff997a043/server/index.js#L73

**Comment by sudheesh001** [comment]
Implemented and reviewd as of 55b22641d6bb11548f6ad7fc559f5c3bd315871d
